mayastor/src/rebuild.rs

```diff
@@ -191,80 +224,115 @@ impl RebuildJob {
             destination_hdl,
             start,
             end,
-            current: start,
+            next: start,
             block_size,
             segment_size_blks,
-            copy_buffer,
+            tasks,
             complete_fn,
             complete_chan: unbounded::<RebuildState>(),
             state: RebuildState::Pending,
         })
     }
 
-    /// Rebuilds a non-healthy child from a healthy child from start to end
+    // Runs the management async task that kicks off N rebuild copy tasks and
+    // awaits each completion before kicking off more tasks until the
+    // bdev is fully rebuilt
     async fn run(&mut self) {
         self.change_state(RebuildState::Running);
-        self.current = self.start;
+        self.next = self.start;
         self.stats();
 
-        while self.current < self.end {
-            if let Err(e) = self.copy_one().await {
-                error!("Failed to copy segment {}", e);
-                self.change_state(RebuildState::Failed);
-                self.send_complete();
-                return;
-            }
-
-            if self.state == RebuildState::Stopped
-                || self.state == RebuildState::Paused
-            {
-                return self.send_complete();
+        self.start_all_tasks();
+        while self.tasks.active > 0 {
+            match self.await_one_task().await {
+                Some(r) => match r.error {
+                    None => {
+                        if self.state == RebuildState::Stopped
+                            || self.state == RebuildState::Paused
+                        {
+                            break;
```

Comment:
Do we need to "await_all_tasks" here? If we break due to a stop or pause event we'll complete the rebuild job even though there may still be active rebuild tasks.

Reply:
oops good catch